File: backend/apps/profiles/subjects.py
"""
Canonical subject-key taxonomy, shared by profile analytics (subject
mastery, skill map) and the subject XP ledger (apps.rankings.SubjectXP).

These 5 keys match apps.mock_exams.MockExamSubject and
apps.flashcards.FlashcardSubject choices exactly. apps.practice.Subject is
free-text content, so it needs its own name mapping rather than a shared
choices field — verify against real Subject rows if new practice content is
added for another subject.
"""

from apps.mock_exams.models import MockExamSubject

SUBJECT_LABELS = {
    "math": "Մաթեմատիկա",
    "physics": "Ֆիզիկա",
    "biology": "Կենսաբանություն",
    "chemistry": "Քիմիա",
    "english": "Անգլերեն",
}

PRACTICE_SUBJECT_NAMES = {
    "math": "Մաթեմատիկա",
    "physics": "Ֆիզիկա",
    "english": "Անգլերեն",
}

_PRACTICE_NAME_TO_KEY = {name: key for key, name in PRACTICE_SUBJECT_NAMES.items()}

# apps.mock_exams.MockExamSubject and apps.flashcards.FlashcardSubject share
# identical (value, English label) pairs, so one reverse map covers both —
# used for MistakeEntry.subject_name snapshots, which store
# card.deck.get_subject_display() / exam.get_subject_display() (English)
# rather than the canonical key itself.
_ENGLISH_LABEL_TO_KEY = {label: key for key, label in MockExamSubject.choices}


def canonical_key_for_practice_subject(subject) -> str | None:
    """apps.practice.Subject instance -> canonical key, or None if that
    subject's name isn't one of the mapped ones (shouldn't happen given
    current content, but content could be added under an unmapped name)."""
    if subject is None:
        return None
    return _PRACTICE_NAME_TO_KEY.get(subject.name)


def subject_key_for_name(name: str) -> str | None:
    """Best-effort reverse lookup from a free-text display name back to the
    canonical key — for data that snapshotted a display string instead of
    the key (e.g. MistakeEntry.subject_name), which can be either an
    Armenian practice subject name or an English MockExamSubject/
    FlashcardSubject label depending on where the mistake came from."""
    if not name:
        return None
    return _PRACTICE_NAME_TO_KEY.get(name) or _ENGLISH_LABEL_TO_KEY.get(name)

File: frontend/src/lib/subjects.ts
import { Atom, BookMarked, Dna, FlaskConical, Languages, Sigma, type LucideIcon } from "lucide-react";

export type SubjectKey = "math" | "physics" | "biology" | "chemistry" | "english";

export interface SubjectMeta {
  key: SubjectKey;
  label: string;
  /*
    A lucide component, not a glyph.

    These were `"∑" | "⚛" | "🧬" | "⚗" | "🇬🇧"` — five characters from three
    different worlds. The first two are mathematical symbols that render in
    the text font at the text colour; the next two are colour emoji that pick
    up per-platform artwork and weight; the last is a *flag*, which stands for
    a country rather than a subject and is the one glyph on the list that some
    platforms refuse to draw at all. Rendered side by side in a subject
    filter, as they were on mock exams, flashcards, groups and the learning
    profile, the row read as five unrelated things.

    One field feeds thirteen call sites across seven surfaces, so this is the
    single highest-leverage place in the codebase to fix the icon language.
  */
  Icon: LucideIcon;
}

export const SUBJECTS: SubjectMeta[] = [
  { key: "math", label: "Մաթեմատիկա", Icon: Sigma },
  { key: "physics", label: "Ֆիզիկա", Icon: Atom },
  { key: "biology", label: "Կենսաբանություն", Icon: Dna },
  { key: "chemistry", label: "Քիմիա", Icon: FlaskConical },
  { key: "english", label: "Անգլերեն", Icon: Languages },
];

/** The icon for a subject named in either language, with a neutral fallback
 *  for anything unmapped — a mistake logged against a subject the frontend
 *  does not know still needs something to draw. */
export function subjectIconForName(name: string | undefined): LucideIcon {
  return subjectMetaForName(name)?.Icon ?? BookMarked;
}

export function subjectMeta(key: string | undefined): SubjectMeta | undefined {
  return SUBJECTS.find((s) => s.key === key);
}

/*
  Some rows snapshot a subject's *English* display label rather than its key:
  MistakeEntry.subject_name stores `exam.get_subject_display()` /
  `deck.get_subject_display()` for mock-exam and flashcard mistakes, which is
  "Mathematics", while practice mistakes store "Մաթեմատիկա". The result is an
  Armenian interface reading "Mathematics · Հանրահաշվական նույնություններ".

  These map a stored name — in either language — back to the canonical
  subject, purely for display. Never use the localized string for filtering or
  links: the API matches on the raw stored value, so a request for
  "Մաթեմատիկա" would return nothing for a mock-exam mistake.
*/
const ENGLISH_SUBJECT_LABELS: Record<string, SubjectKey> = {
  Mathematics: "math",
  Physics: "physics",
  Biology: "biology",
  Chemistry: "chemistry",
  English: "english",
};

export function subjectMetaForName(name: string | undefined): SubjectMeta | undefined {
  if (!name) return undefined;
  const byArmenian = SUBJECTS.find((s) => s.label === name);
  if (byArmenian) return byArmenian;
  const key = ENGLISH_SUBJECT_LABELS[name];
  return key ? SUBJECTS.find((s) => s.key === key) : undefined;
}

/** Display-only: the Armenian label for a stored subject name, falling back to
 *  the raw name for anything unmapped. */
export function localizeSubjectName(name: string | undefined): string {
  return subjectMetaForName(name)?.label ?? name ?? "";
}

// Maps a subject key to the practice.Subject.name it corresponds to, for the
// subjects that already have topic-tree practice content. Keys with no entry
// here don't have practice content yet — the hub shows "coming soon" instead.
export const PRACTICE_SUBJECT_NAMES: Partial<Record<SubjectKey, string>> = {
  math: "Մաթեմատիկա",
  english: "Անգլերեն",
  physics: "Ֆիզիկա",
};
